File: sdk/python/examples/quickstart.py
"""Fetch a profile and the most recent scrobbles for a handle.

Run:
    uv run python examples/quickstart.py alice.bsky.social
"""

from __future__ import annotations

import asyncio
import sys

from rocksky import Client, NotFoundError


async def main(handle: str) -> int:
    async with Client() as rocksky:
        try:
            profile = await rocksky.actor.get_profile(handle)
        except NotFoundError:
            print(f"no profile for {handle!r}")
            return 1

        print(f"{profile.display_name or profile.handle}  ({profile.did})")
        print(f"  joined: {profile.created_at}")
        print()

        scrobbles = await rocksky.scrobble.list(did=profile.did, limit=10)
        if not scrobbles:
            print("(no scrobbles yet)")
            return 0

        print("recent scrobbles:")
        for s in scrobbles:
            stamp = s.date.isoformat() if s.date else "—"
            print(f"  [{stamp}]  {s.artist} — {s.title}")
        return 0


if __name__ == "__main__":
    handle = sys.argv[1] if len(sys.argv) > 1 else "tsiry.dev"
    sys.exit(asyncio.run(main(handle)))

File: sdk/rust/examples/quickstart.rs
//! Fetch a profile and recent scrobbles.
//!
//! ```text
//! cargo run --example quickstart -- alice.bsky.social
//! ```

use rocksky::{Client, Error};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let handle = std::env::args().nth(1).unwrap_or_else(|| "tsiry.dev".into());

    let client = Client::new();

    let profile = match client.actor().get_profile(&handle).await {
        Ok(p) => p,
        Err(e) if e.is_not_found() => {
            eprintln!("no profile for {handle:?}");
            std::process::exit(1);
        }
        Err(Error::Api { status: 400, .. }) => {
            eprintln!("invalid handle {handle:?}");
            std::process::exit(1);
        }
        Err(e) => return Err(e.into()),
    };

    println!(
        "{}  ({})",
        profile
            .display_name
            .clone()
            .or_else(|| profile.handle.clone())
            .unwrap_or_else(|| "—".into()),
        profile.did.clone().unwrap_or_else(|| "—".into())
    );
    if let Some(ts) = profile.created_at {
        println!("  joined: {ts}");
    }
    println!();

    let Some(did) = profile.did.clone() else {
        eprintln!("(profile has no DID, can't list scrobbles)");
        return Ok(());
    };

    let scrobbles = client.scrobble().list().did(did).limit(10).send().await?;
    if scrobbles.is_empty() {
        println!("(no scrobbles yet)");
        return Ok(());
    }

    println!("recent scrobbles:");
    for s in scrobbles {
        let when = s
            .date
            .map(|d| d.to_rfc3339())
            .unwrap_or_else(|| "—".into());
        println!(
            "  [{when}]  {} — {}",
            s.artist.unwrap_or_else(|| "?".into()),
            s.title.unwrap_or_else(|| "?".into())
        );
    }
    Ok(())
}
